contracts/cw3-flex-multisig/src/state.rs

```diff
@@ -91,15 +126,17 @@ impl<'a> IndexList<Proposal> for ProposalIndexes<'a> {
 }
 
 /// Returns a value that can be used as a secondary index key in the proposals map
-pub fn status_index(status: Status) -> Vec<u8> {
-    vec![status as u8]
+pub fn status_height_index(status: Status, height: u64) -> Vec<u8> {
+    let mut idx = vec![status as u8];
+    idx.extend_from_slice(&height.to_be_bytes());
+    idx
 }
 
 // secondary indexes on state for PROPOSALS to find all open proposals efficiently
 pub fn proposals<'a>() -> IndexedMap<'a, U64Key, Proposal, ProposalIndexes<'a>> {
     let indexes = ProposalIndexes {
         status: MultiIndex::new(
-            |p| status_index(p.status),
+            |p| status_height_index(p.status, p.start_height),
```

Comment:
If I understand correctly, you want to query all Open proposals after a given height, right? So you will know which proposals need to use the new weight for an addr, vs. the old one.

Maybe I'm missing something, but, why not do it the other way around?

When a group update for an addr comes (which you are subscribed, because of the hook), store the addr (as key) and the current block height, along with the original weight (values) in a map. Something like original_weights or so. All this is O(1).

When voting, check if the voter is in the original_weights map (O(1)), and if it is, check if the stored block height is greater than the start_height of the proposal. If that's the case, use the stored original weight (i.e. ignore the update).
In any other case, query the group contract for the weight.

Makes sense?

Reply:
What I'm probably missing is: multiple updates over the same addr.

In that case, it would be good to use an index, `original_weights` per `block_height`.
That way, we'll have one index per (updated) addr, where we can efficiently query the original value before a given block height (which would be the `start_height` of each Open proposal).

Given that block heights are monotonically increasing, an array and binary search will do. But an index is probably better / nicer.

Reply:
Update: I see now that this functionality in the index is just an optimization, to avoid creating the snapshots if there is no currently open proposal. Cool.

Reply:
Yup, I could write every update, but if we have token weighted voting, a stake may change 30 times during an open proposal,and it will only store the last one